### PR TITLE
Feature/437 make auth method configurable

### PR DESCRIPTION
**Pull request by lokm01**
This PR allows configuration of menas API authentication method. Currently available are kerberos and inmemory

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala [hunk 1]
```diff
@@ -137,76 +67,82 @@ class WebSecurityConfig {
...
 
   @Component
   class RestAuthenticationEntryPoint extends AuthenticationEntryPoint {
     override def commence(request: HttpServletRequest,
-                          response: HttpServletResponse,
-                          authException: AuthenticationException): Unit = {
-      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized")
+        response: HttpServletResponse,
+        authException: AuthenticationException): Unit = {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, s"Unauthorized")
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala]: Why is there an `s` before plain string?

File: menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  @Value("${za.co.absa.enceladus.menas.auth.sysuser.keytab.location:}")
+  val keytabLocation: String = ""
+  @Value("${za.co.absa.enceladus.menas.auth.ldap.search.base:}")
+  val ldapSearchBase: String = ""
+  @Value("${za.co.absa.enceladus.menas.auth.ldap.search.filter:}")
+  val ldapSearchFilter: String = ""
+  @Value("${za.co.absa.enceladus.menas.auth.kerberos.debug:false}")
+  val kerberosDebug: Boolean = false
+
+  lazy val requiredParameters = Seq((adDomain, "za.co.absa.enceladus.menas.auth.ad.domain"),
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala]: either annotate if you expect to use it somewhere else or make private.

File: menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    (adServer, "za.co.absa.enceladus.menas.auth.ad.server"),
+    (servicePrincipal, "za.co.absa.enceladus.menas.auth.sysuser.principal"),
+    (keytabLocation, "za.co.absa.enceladus.menas.auth.sysuser.keytab.location"),
+    (ldapSearchBase, "za.co.absa.enceladus.menas.auth.ldap.search.base"),
+    (ldapSearchFilter, "za.co.absa.enceladus.menas.auth.ldap.search.filter"))
+
+  override def afterPropertiesSet() {
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala]:
```suggestion
  override def afterPropertiesSet() = {
```

You should always include `=` by the [style guide](https://docs.scala-lang.org/style/declarations.html#procedure_syntax)

File: menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    System.setProperty("javax.net.debug", kerberosDebug.toString)
+    System.setProperty("sun.security.krb5.debug", kerberosDebug.toString)
+  }
+
+  private def validateParam(param: String, paramName: String) = {
+    if (param.isEmpty()) throw new IllegalArgumentException(s"$paramName has to be configured in order to use kerberos Menas authentication")
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala]:
```suggestion
    if (param.isEmpty()) {
      throw new IllegalArgumentException(s"$paramName has to be configured in order to use kerberos Menas authentication")
    }
```
Subjectively better.

File: menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  private def kerberosServiceAuthenticationProvider() = {
+    val provider = new KerberosServiceAuthenticationProvider()
+    provider.setTicketValidator(sunJaasKerberosTicketValidator())
+    provider.setUserDetailsService(ldapUserDetailsService())
+    provider
+  }
+
+  override def configure(auth: AuthenticationManagerBuilder) {
+    this.validateParams()
+    auth.authenticationProvider(kerberosServiceAuthenticationProvider()).authenticationProvider(activeDirectoryLdapAuthenticationProvider())
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/auth/KerberosMenasAuthentication.scala]:
```suggestion
   auth.authenticationProvider(kerberosServiceAuthenticationProvider())
      .authenticationProvider(activeDirectoryLdapAuthenticationProvider())
```

As per the agreement from the CQC

File: menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala [hunk 6]
```diff
@@ -137,76 +67,82 @@ class WebSecurityConfig {
 
   @Configuration
   @Order(1)
-  class ApiWebSecurityConfigurationAdapter @Autowired()(restAuthenticationEntyPoint: RestAuthenticationEntryPoint,
-                                                        authenticationSuccessHandler: AuthSuccessHandler,
-                                                        authenticationFailureHandler: AuthenticationFailureHandler,
-                                                        logoutSuccessHandler: LogoutSuccessHandler)
+  class ApiWebSecurityConfigurationAdapter @Autowired() (restAuthenticationEntyPoint: RestAuthenticationEntryPoint,
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala]:
```suggestion
  class ApiWebSecurityConfigurationAdapter @Autowired() (restAuthenticationEntryPoint: RestAuthenticationEntryPoint,
```

And all the other occurrences.

File: menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala [hunk 7]
```diff
@@ -137,76 +67,82 @@ class WebSecurityConfig {
...
+      authMechanism.toLowerCase match {
+        case "inmemory" => {
+          logger.info("Using InMemory Menas authentication")
+          beanFactory.getBean(classOf[InMemoryMenasAuthentication])
+        }
+        case "kerberos" => {
+          logger.info("Using Kerberos Menas Authentication")
+          beanFactory.getBean(classOf[KerberosMenasAuthentication])
+        }
+        case _ => throw new IllegalArgumentException(s"Invalid authentication mechanism - use one of: inmemory, kerberos")
```
Comment (Zejnilovic) [inline menas/src/main/scala/za/co/absa/enceladus/menas/WebSecurityConfig.scala]: Why not also have this on the new line as previous matches.

And brackets are redundant here.

File: menas/src/main/resources/application.properties.template [hunk 8]
```diff
@@ -20,12 +20,19 @@ multipart.maxFileSize=16Mb
 # Set the total request size for a multipart/form-data (default 10Mb)
 multipart.maxRequestSize=16Mb
 
+# How will users authenticate to menas. Available options: inmemory, kerberos
+za.co.absa.enceladus.menas.auth.mechanism = kerberos
```
Comment (Zejnilovic) [inline menas/src/main/resources/application.properties.template]: I don't like making Kerberos default. I think that we should have inmemory default as that is the "works out of the box" solution.

**Comment by lokm01** [comment]
Fair enough...

Though I think we're putting little too much effort on reviewing things which in my opinion are questions of taste.

File: menas/src/main/resources/application.properties.template [hunk 9]
```diff
@@ -21,7 +21,7 @@ multipart.maxFileSize=16Mb
 multipart.maxRequestSize=16Mb
 
 # How will users authenticate to menas. Available options: inmemory, kerberos
-za.co.absa.enceladus.menas.auth.mechanism = kerberos
+za.co.absa.enceladus.menas.auth.mechanism = inmemory
```
Comment (Zejnilovic) [inline menas/src/main/resources/application.properties.template]: You would need to uncomment lines `34` and `35`

**Pull request merged**